plat/mediatek/mt8183/include/plat_macros.S

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2014-2015, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <cci.h>
+#include <gic_common.h>
+#include <gicv2.h>
+#include <platform_def.h>
+
+.section .rodata.gic_reg_name, "aS"
+gicc_regs:
+	.asciz "gicc_hppir", "gicc_ahppir", "gicc_ctlr", ""
+gicd_pend_reg:
+	.asciz "gicd_ispendr regs (Offsets 0x200 - 0x278)\n"	\
+		" Offset:\t\t\tvalue\n"
+newline:
+	.asciz "\n"
+spacer:
+	.asciz ":\t\t0x"
+
+.section .rodata.cci_reg_name, "aS"
+cci_iface_regs:
+	.asciz "cci_snoop_ctrl_cluster0", "cci_snoop_ctrl_cluster1" , ""
+
+	/* ---------------------------------------------
+	 * The below macro prints out relevant GIC and
+	 * CCI registers whenever an unhandled exception
+	 * is taken in BL3-1.
+	 * Clobbers: x0 - x10, x16, x17, sp
+	 * ---------------------------------------------
+	 */
+	.macro plat_crash_print_regs
+	mov_imm x16, BASE_GICD_BASE
+	mov_imm x17, BASE_GICC_BASE
```

Comment:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
This is copy-pasted from Arm platforms code. If you take a look at the macros there it explicitly says that x16 and x17 are used to pass the base addresses. The registers aren't used as a callee-saved registers, so it is fine. However, I agree that it isn't a bad idea to change the registers, using x26 and x27 makes the code more intuitive.

Reply:
Oh, ok, I see what you mean. Sure, if you are using the multi console for crash output then you need to use callee-saved registers, I didn't realize about that problem because only rockchip platforms do that, as far as I know. However, will it be enough to move x16 and x17? I haven't checked all the code in this macro, but I'd be surprised if x16 and x17 were the only problems here.

Reply:
In general we probably have to do a cleanup of all assembly code and make it AAPCS compliant. I think that in the crash handling code we didn't do it because we never assumed that we would need to use more than really simple functions after a crash.

Reply:
Yeah, you're right, the code has more problems than that. I think the rest of it barely manages to scoot around conflicts. I tried to pick the extra registers for multi-console carefully so that most code that was written for the existing crash handlers should still work (i.e. taking the extra registers from the back of the caller-saved block, x15, x14, x13, which are less likely to have already been used).

But the other problem is (if I understand the ABI right), x16 and x17 are reserved for the linker and could get clobbered at any BL instruction, regardless of the code on both sides. The linker is always allowed to automatically insert a veneer in the call which clobbers these registers. Maybe the situations where veneers are needed just don't happen inside TF in practice, but the ABI doesn't specify them so it still seems like a very bad idea to trust these to hold their value across branch instructions ever.

(It's also not great that this code has been copied and mangled across so many platforms by now... it would really be better to maintain a single version in include/drivers/arm/gic_helpers.S or something like that.)

Reply:
> In general we probably have to do a cleanup of all assembly code and make it AAPCS compliant.

Just to clarify, AAPCS is only possible where you have a backing store (stack) available when functions calls are nested more than a level. Since there are many places (like crash reporting, early init code) where stack is not available/reliable, we have rely on custom register clobber list. 

> Some parts of the multi-console core explicitly make use of these as scratch registers, so that would definitely blow up.

I see that the console_cbmc_putc() uses x16. The crash console framework had the assumption that the console_core_* functions would only use a couple of registers but MULTI-CONSOLE changed that.

Usually, when not following AAPCS, the convention is that the top level functions use the higher registers while the inner functions works on a smaller subset with the leaf function using the minimal and lowest set of registers.  WIth MULTI-CONSOLE, this is no longer true and, as you say, it is very difficult to track which register to use.

I think it is about time the crash framework is re-implemented for the MULTI-CONSOLE so that x0-x17 can be used by multi-console drivers. I will create a tf-issue for this.

> But the other problem is (if I understand the ABI right), x16 and x17 are reserved for the linker and could get clobbered at any BL instruction, regardless of the code on both sides

I agree it is not safe to use these registers across function calls and should be changed. Factoring out the useful bits in ARM platform to comon code is something we have started doing and this is another bit that can be in common code.

Reply:
> Just to clarify, AAPCS is only possible where you have a backing store (stack) available when functions calls are nested more than a level.

I wouldn't completely agree with that. What you're saying is true if you want every individual function to be AAPCS compliant, which we just can't really do if we want a console framework that doesn't require a stack. But you can still have an API that is AAPCS compliant from the outside even though internally it is implemented in assembler and relying on custom ABIs between its individual functions. MULTI_CONSOLE works that way (and technically, the old console framework already worked that way too), and it has to because all the API functions can be called from either assembly or C runtime contexts. For example, the normal C function putchar() calls the assembly function console_putc(), so the latter needs to be AAPCS compliant on the outside for this to work.

> I think it is about time the crash framework is re-implemented for the MULTI-CONSOLE so that x0-x17 can be used by multi-console drivers. I will create a tf-issue for this.

Not quite sure what you mean by that. Are you suggesting that MULTI_CONSOLE framework functions should be using higher registers (e.g. x29) instead? Since it has no stack to save them, this would break calls from normal C functions as described above, so that seems like a bad idea.

Actually, I just noticed that someone has already done this: https://github.com/ARM-software/arm-trusted-firmware/commit/d35cc347. Unfortunately I didn't notice that change when it went in. This presumably breaks all MULTI_CONSOLE crash reporting (do_crash_reporting uses SP as a scratch register) which was one of the main use cases I implemented the framework for, so I'd prefer if we could undo that.

Reply:
The problem with undoing that commit is that now there is one console driver written in C (in a AArch32 port, not AArch64). The console framework needs to use AAPCS at least in AArch32, so it would be unexpected if it didn't in AArch64.

https://github.com/ARM-software/arm-trusted-firmware/blob/9c10e3485b10ece95f9af8a6af79558397ee1d78/drivers/imx/uart/imx_uart.c

Reply:
> I wouldn't completely agree with that. What you're saying is true if you want every individual function to be AAPCS compliant, which we just can't really do if we want a console framework that doesn't require a stack.  But you can still have an API that is AAPCS compliant from the outside even though internally it is implemented in assembler and relying on custom ABIs between its individual functions.

I agree. That's what I meant. Not all the assembly functions can be made AAPCS. As long as the APIs exported are AAPCS, it can be invoked from C.

> Not quite sure what you mean by that. Are you suggesting that MULTI_CONSOLE framework functions should be using higher registers (e.g. x29) instead? Since it has no stack to save them, this would break calls from normal C functions as described above, so that seems like a bad idea.

What I meant is rework the to crash reporting to use the higher register set, and the MULTI_CONSOLE stays as it is. The MULTI_CONSOLE API should remain AAPCS compliant.

> Actually, I just noticed that someone has already done this: d35cc34. Unfortunately I didn't notice that change when it went in.

Sorry, didn't realise that. Unfortunately interaction between MULTI_CONSOLE and crash reporting is something that I didn't understand too well till recent. Since we override the plat_crash_console_* functions, we couldn't detect the problem on ARM reference platforms.

It is a an useful capability to allow platforms to implement their console drivers in C, as long as they don't use it for crash reporting which is what the patch is allowing to do.

@antonio-nino-diaz-arm and I have had a discussion and we think the solution here is to have a new API `crash_console_putc` which is equivalent `console_putc` but doesn't use the stack. The default `plat_crash_console_putc` will call into `crash_console_putc`.   The implementation of `crash_console_putc` can be simpler than `console_putc` because it doesn't need to check for the 
`console_state` as it is always invoked in the crash state.  This means the `plat_crash_console_init` can be empty. Similarly `crash_console_flush` can be implemented.

Please let us know of your opinion.